src/snowcli/cli/nativeapp/version/version_processor.py

```diff
@@ -0,0 +1,306 @@
+import logging
+from pathlib import Path
+from textwrap import dedent
+from typing import Dict, List, Optional
+
+import typer
+from click import ClickException
+from git import Repo
+from git.exc import InvalidGitRepositoryError
+from rich import print
+from snowcli.cli.nativeapp.artifacts import find_version_info_in_manifest_file
+from snowcli.cli.nativeapp.constants import VERSION_COL
+from snowcli.cli.nativeapp.exceptions import ApplicationPackageDoesNotExistError
+from snowcli.cli.nativeapp.manager import (
+    NativeAppCommandProcessor,
+    NativeAppManager,
+    ensure_correct_owner,
+)
+from snowcli.cli.nativeapp.policy import PolicyBase
+from snowcli.cli.nativeapp.run_processor import NativeAppRunProcessor
+from snowcli.cli.nativeapp.utils import (
+    find_all_rows,
+    find_first_row,
+)
+from snowcli.cli.project.util import unquote_identifier
+from snowcli.exception import SnowflakeSQLExecutionError
+from snowflake.connector import ProgrammingError
+from snowflake.connector.cursor import DictCursor
+
+log = logging.getLogger(__name__)
+
+
+def check_index_changes_in_git_repo(
+    project_root: Path,
+    policy: PolicyBase,
+) -> None:
+    """
+    Checks if the project root, i.e. the native apps project is a git repository. If it is a git repository,
+    it also checks if there any local changes to the directory that may not be on the app package stage.
+    """
+    try:
+        repo = Repo(project_root, search_parent_directories=True)
+        assert repo.git_dir is not None
+
+        # Check if the repo has any changes, including untracked files
+        if repo.is_dirty(untracked_files=True):
+            print("Changes detected in your git repository!")
+            repo.git.execute(["git", "status"])
+
+            user_prompt = (
+                "You have local changes in this repository that are not part of a previous commit. Do you still want to continue?",
+            )
+            if not policy.should_proceed(user_prompt):
+                if policy.exit_code:
+                    print(
+                        "Cannot create a new version non-interactively without --force."
+                    )
+                else:
+                    print("Not creating a new version.")
```

Comment:
From a responsibility POV, this seems backwards. This logic knows whether being interactive is a dealbreaker. The policy shouldn't. Otherwise we get some unwanted coupling that make policies difficult to evolve/reuse in different contexts. I think this should read something like:

```
should_proceed = policy...
if not should_proceed:
    if is_interactive: # we'll need to carry this context around
        raise SomeError("Cannot create a new version ...")
    else:
        raise SomeError("Not creating a new version")
```

We can either thread the interactive state through the processors, or query the policies for it if you prefer. I'm fine either way.